docs/developer/python-openflow.rst

```diff
@@ -0,0 +1,206 @@
+python-openflow
+***************
+
+Basic lib organization
+======================
+
+Repository file tree
+++++++++++++++++++++
+
+::
+
+  - docs/
+  - pyof/
+    - foundation/
+    - v0x01/
+    - v0x04/
+  - raw/
+  - tests/
+
+Description
++++++++++++
+
+- docs/
+    Sphinx documentation.
+
+- pyof/
+    Main python-openflow module.
+
+- pyof/foundation/
+    Subpackage containing classes used/needed to implement a new pyof message
+    or structure, like ``foundation.base.GenericStruct`` and many basic types
+    like the null terminated string ``foundation.basic_types.Char``.
+
+- pyof/v0x01/
+    Subpackage containing OpenFlow version 1.0 (0x01) classes for messages and
+    structs.
+
+- pyof/v0x04/
+    Subpackage containing OpenFlow version 1.3 (0x04) classes for messages and
+    structs.
+
+- raw/
+    Raw OpenFlow message dumps to be used by the test suite.
+
+- tests/
+    Package with python-openflow test suite.
+
+
+How to use the python-openflow library
+======================================
+
+importing
++++++++++
+To use the python-openflow module you need to import its classes directly
+from its submodule.
+
+how to do it
+~~~~~~~~~~~~
+To create a v0x04 Hello message for example, you need to import its class
+from the ``pyof.v0x04.symmetric.hello`` submodule using the statement:
+
+.. code:: python
+
+  >>> from pyof.v0x01.symmetric.hello import Hello
+  >>> hello_message = Hello()
+
+how NOT to do it
+~~~~~~~~~~~~~~~~
+Importing only ``pyof``, a version subpackage like ``pyof.v0x01``,
+or even a message category subpackage like ``pyof.v0x01.symmetric`` doesn't
+provide you anything.
+
+.. code:: python
+
+  >>> import pyof
+  >>> pyof.v0x01.symmetric.hello.Hello()
+
+  AttributeError: module 'pyof' has no attribute 'v0x01'
+
+
+.. code:: python
+
+  >>> old_dir = dir() + ['old_dir']
+
+  >>> from pyof import *
+  >>> from pyof.v0x01 import *
+  >>> from pyof.v0x01.symmetric import *
+
+  >>> [i for i in dir() if i not in old_dir]
+  []
+
+nothing is imported...
+
+pyof Objects
+++++++++++++
+pyof objects are instances of GenericStruct or GenericType or other classes
+derived from these.
+They provide the methods ``pack`` ``unpack`` and ``get_size``.
+
+initialization
+++++++++++++++
+``GenericType``s accept as argument an initial value.
+``GenericStruct``s derivatives usually accept as argument initial values
+for its class attributes.
+
+packing
++++++++
+To pack a pyof object, simply call its ``pack()`` method, which returns a
+``bytes`` object of the binary representation.
+
+unpacking
++++++++++
+To unpack a buffer as pyof object, simply initialize a new object and
+call its ``unpack()`` method, passing the buffer as the argument. The
+buffer will be unpacked in place setting the attributes of the object
+instance used to make the call.
```

Comment:
This does not work for "Messages", where the unpack receives the binary data without the message header.